helper/resource/id.go

```diff
@@ -27,14 +28,35 @@ func UniqueId() string {
 // across multiple terraform executions, as long as the clock is not turned back
 // between calls, and as long as any given terraform execution generates fewer
 // than 4 billion IDs.
-func PrefixedUniqueId(prefix string) string {
+func PrefixedUniqueId(prefix string, max_length ...int) string {
 	// Be precise to 4 digits of fractional seconds, but remove the dot before the
 	// fractional seconds.
 	timestamp := strings.Replace(
 		time.Now().UTC().Format("20060102150405.0000"), ".", "", 1)
 
+	// If max_length is provided, truncate the prefix to that size, otherwise use
+	// the prefix provided
+	var final_prefix string
+	if len(max_length) > 0 {
+		final_prefix = TruncatedPrefix(prefix, max_length[0])
+	} else {
+		final_prefix = prefix
+	}
+
 	idMutex.Lock()
 	defer idMutex.Unlock()
 	idCounter++
-	return fmt.Sprintf("%s%s%08x", prefix, timestamp, idCounter)
+	return fmt.Sprintf("%s%s%08x", final_prefix, timestamp, idCounter)
+}
+
+func TruncatedPrefix(prefix string, max_length int) string {
```

Comment:
Given that `TruncatedPrefix` and `MaxInt` are used only once, and they are not long functions, I'd prefer to inline them in the main function just to avoid these being in the global scope where their context is not obvious. After inlining I think we can also avoid the conversion to/from floats with `math.Max`, since we know the lower bound will always be zero:

```
if maxLength > len(prefix) {
  maxLength = len(prefix)
}
```

Note that in Go the first letter being capitalized signals that a function is part of the public API of a package, so for internal helper functions you'd normally name them with a lowercase letter to mark them as private to the package.